arkouda/apply.py

```diff
@@ -0,0 +1,99 @@
+from typeguard import typechecked
+from typing import Callable, Union, Optional
+
+from arkouda.client import generic_msg
+from arkouda.pdarrayclass import pdarray
+from arkouda.pdarraycreation import create_pdarray
+import cloudpickle
+import base64
+import numpy as np
+from arkouda.numpy.dtypes import dtype
+
+
+__all__ = [
+    "apply",
+]
+
+
+# TODO: it would be nice to typecheck that func takes and returns arr.dtype,
+# but that likely requires making pdarray generic over its dtype
+@typechecked
+def apply(
+    arr: pdarray,
+    func: Union[Callable, str],
+    result_dtype: Optional[Union[np.dtype, str]] = None,
+) -> pdarray:
+    """
+    Apply a python function to a pdarray. The function should take 1 argument
+    and return a new value. The function will then be called on each element in
+    the pdarray.
+
+    For example,
+    >>> apply(ak.array([1, 2, 3]), lambda x: x+1)
+
+    Or,
+    >>> import math
+    >>> arr = ak.randint(0, 10, 10_000)
+    >>> def times_pi(x):
+            return x*math.pi
+    >>> ak.apply(arr, times_pi, "float64")
+
+    Warning: This function is experimental and may not work as expected.
+    Known limitations:
+    - The use of some python modules inside of the function may crash the server.
+    Known modules know to crash the server are:
+        - numpy
+        - pandas
+        - matplotlib
+    - Any python modules used inside of the function must be installed on the server.
+
+    Parameters
+    ----------
+    arr : pdarray
+        The pdarray to which the function is applied
+
+    func : Union[Callable, str]
+
+        The function to apply to the array. This can be a callable function or
+        a string, but either way it should take a single argument and return a
+        single value. If a string, it should be a lambda function that takes a
+        single argument, e.g. "lambda x,: x+1".
```

Comment:
I remember dangling comma being necessary for a reason I can't recall. Is that the case? If so, could you add that somewhere here?